From 261c0ce6723d19bd41ee1745ad51fb7d88779428 Mon Sep 17 00:00:00 2001
From: coldhex <coldhex@localhost>
Date: Sat, 19 Jul 2025 21:43:26 +0300
Subject: [PATCH] nv2a: Fix split element index arrays due to repeated Begin

The game "Star Wars Jedi Knight: Jedi Academy" repeats push Begin in the
middle of sending element indices for vertices of triangle lists, which
causes severe artifacts if treated as separate triangle lists. This commit
ignores the repeated Begin.
---
 hw/xbox/nv2a/pgraph/pgraph.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/hw/xbox/nv2a/pgraph/pgraph.c b/hw/xbox/nv2a/pgraph/pgraph.c
index 98f4f3a880..a316833d13 100644
--- a/hw/xbox/nv2a/pgraph/pgraph.c
+++ b/hw/xbox/nv2a/pgraph/pgraph.c
@@ -2529,6 +2529,20 @@ DEF_METHOD(NV097, SET_BEGIN_END)
         pgraph_reset_inline_buffers(pg);
         pg->primitive_mode = PRIM_TYPE_INVALID;
     } else {
+        if (pg->primitive_mode != PRIM_TYPE_INVALID &&
+            pg->primitive_mode == parameter) {
+            /* The game "Star Wars Jedi Knight: Jedi Academy" sends Begin in the
+             * middle of sending ARRAY_ELEMENT16 values when primitive_mode is
+             * PRIM_TYPE_TRIANGLES, splitting a triangle (e.g. having one vertex
+             * before begin and two after) and even worse, the triangle vertices
+             * after the Begin are then out of sync leading to severe artifacts.
+             * Before the second Begin there are 511 (=0x1FF) ARRAY_ELEMENT16
+             * values so perhaps a new Begin is required by NV2A after that
+             * amount.
+             */
+            NV2A_DPRINTF("Begin without End, same primitive mode!\n");
+            return;
+        }
         if (pg->primitive_mode != PRIM_TYPE_INVALID) {
             NV2A_DPRINTF("Begin without End!\n");
         }
